### Gaussian Kernel Regression


In [1]:
# load dependencies'
import pandas as pd
pd.set_option('display.max_columns', None)# pandas show all columns of table instead of restricted#
pd.options.mode.chained_assignment = None  # default='warn' surpresses warnings at spatial distribution
import geopandas as gpd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from scipy import stats
import numpy as np
import math
import plotly.express as px
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error


from envirocar import TrackAPI, DownloadClient, BboxSelector, ECConfig
from envirocar import correction as correct
from envirocar import inspection as inspect
from envirocar import manipulation as manipulate
from envirocar import GKR


# create an initial but optional config and an api client
config = ECConfig()
track_api = TrackAPI(api_client=DownloadClient(config=config))


In [3]:
bbox = BboxSelector([
    7.554130554199218, # min_x
    51.95590322041212, # min_y
    7.590351104736328, # max_x
    51.97874790276371  # max_y
])

# issue a query
track_df = track_api.get_tracks(bbox=bbox, num_results=2) 

id                       time  \
0   5f0ef89c00375c5a2641ef86  2020-07-15T12:37:03+00:00   
1   5f0ef89c00375c5a2641ef88  2020-07-15T12:37:09+00:00   
2   5f0ef89c00375c5a2641ef89  2020-07-15T12:37:14+00:00   
3   5f0ef89c00375c5a2641ef8a  2020-07-15T12:37:19+00:00   
4   5f0ef89c00375c5a2641ef8b  2020-07-15T12:37:24+00:00   
5   5f0ef89c00375c5a2641ef8c  2020-07-15T12:37:29+00:00   
6   5f0ef89c00375c5a2641ef8d  2020-07-15T12:37:34+00:00   
7   5f0ef89c00375c5a2641ef8e  2020-07-15T12:37:39+00:00   
8   5f0ef89c00375c5a2641ef8f  2020-07-15T12:37:44+00:00   
9   5f0ef89c00375c5a2641ef90  2020-07-15T12:37:49+00:00   
10  5f0ef89c00375c5a2641ef91  2020-07-15T12:37:54+00:00   
11  5f0ef89c00375c5a2641ef92  2020-07-15T12:37:59+00:00   
12  5f0ef89c00375c5a2641ef93  2020-07-15T12:38:05+00:00   
13  5f0ef89c00375c5a2641ef94  2020-07-15T12:38:10+00:00   
14  5f0ef89c00375c5a2641ef95  2020-07-15T12:38:15+00:00   
15  5f0ef89c00375c5a2641ef96  2020-07-15T12:38:20+00:00   
16  5f0ef89c00375c5a2641ef97  2020-07-15T12:38:25+00:00   
0   5f0efab600375c5a2642243b  2020-07-15T12:35:22+00:00   
1   5f0efab600375c5a2642243d  2020-07-15T12:35:27+00:00   
2   5f0efab600375c5a2642243e  2020-07-15T12:35:32+00:00   
3   5f0efab600375c5a2642243f  2020-07-15T12:35:37+00:00   

                    geometry  Engine Load.value Engine Load.unit  \
0   POINT (7.57939 51.96766)          30.459892                %   
1   POINT (7.57955 51.96757)          49.230105                %   
2   POINT (7.57988 51.96740)          78.649652                %   
3   POINT (7.58049 51.96715)          31.200400                %   
4   POINT (7.58107 51.96682)          29.735773                %   
5   POINT (7.58154 51.96654)          62.537844                %   
6   POINT (7.58205 51.96621)          35.909973                %   
7   POINT (7.58258 51.96582)          28.092692                %   
8   POINT (7.58307 51.96541)          31.254127                %   
9   POINT (7.58373 51.96520)          39.308760                %   
10  POINT (7.58465 51.96533)          52.792462                %   
11  POINT (7.58563 51.96543)          25.505112                %   
12  POINT (7.58661 51.96548)          30.588237                %   
13  POINT (7.58718 51.96543)          30.588235                %   
14  POINT (7.58721 51.96542)           2.376463                %   
15  POINT (7.58700 51.96541)           0.000000                %   
16  POINT (7.58700 51.96541)                NaN              NaN   
0   POINT (7.57498 51.97452)          28.681476                %   
1   POINT (7.57552 51.97412)          26.665718                %   
2   POINT (7.57595 51.97376)          30.499422                %   
3   POINT (7.57606 51.97343)          58.107638                %   

    Calculated MAF.value Calculated MAF.unit  Speed.value Speed.unit  \
0               3.113889                 g/s    15.978930       km/h   
1               9.778811                 g/s    11.134565       km/h   
2              25.066406                 g/s    33.976330       km/h   
3               7.419664                 g/s    39.000000       km/h   
4               3.106679                 g/s    34.321667       km/h   
5               9.954607                 g/s    30.291912       km/h   
6               8.377679                 g/s    40.000001       km/h   
7               3.479236                 g/s    42.127695       km/h   
8               3.914605                 g/s    37.932298       km/h   
9               8.327728                 g/s    44.000000       km/h   
10             11.645264                 g/s    43.937871       km/h   
11              5.723821                 g/s    46.999999       km/h   
12              3.206776                 g/s    36.840532       km/h   
13              3.103614                 g/s     7.662095       km/h   
14              0.000000                 g/s     0.000000       km/h   
15              0.000000                 g/s     0.000000       km/h   
16    

In [ ]:
correct.drop_dublicates(track_df)

In [ ]:
# take only numerical variables
track_df_numeric = manipulate.get_numerical(track_df)

Inspect missing values to choose a variable which has many missing values. 
We will try to impute CO2 Emission (GPS-based).value

In [ ]:
#missingValues=inspect.sum_missing_values(track_df_numeric)
missingValues=inspect.missing_values_per_variable(track_df_numeric, dropCol=True)
missingValues

Just to get an impression, chose the variable which has the strongest parametric relationship with CO2 Emission (GPS-based).value
here it seems to be he Speed.value
So we will try to impute CO2 Emission (GPS-based).value based on Speed.value

In [ ]:
allCoeffs, very_strong, strong, moderate, weak = inspect.get_classified_correlations(track_df_numeric, 'spearman')
allCoeffs.loc[(allCoeffs['column'] == 'Consumption (GPS-based).value')]

In [ ]:
# get impression of the two variables you want to relate to in order to get 
relation = track_df[["track.id","Speed.value", "CO2 Emission (GPS-based).value"]]
correct.flag_outlier_in_sample(relation, dropOutlierColumn=True, setOutlierToNan=True, dropFlag=True)
relation

In [ ]:
fig = px.scatter(relation, x="Speed.value", y="CO2 Emission (GPS-based).value")
fig.show()

In [ ]:
# drop all rows which contain NaN
relation2 = relation.dropna()
relation2.reset_index(drop=True, inplace=True)
relation2

In [ ]:
inspect.plot_linear_regression(relation2["Speed.value"], relation2["CO2 Emission (GPS-based).value"])

In [ ]:
gaussianKernelRegression=GKR(relation2['Speed.value'],relation2['CO2 Emission (GPS-based).value'], 10)
gaussianKernelRegression.visualize_kernels(100)

In [ ]:
# Predict single value
gaussianKernelRegression.predict(15.978929817676544)

In [ ]:
predictedDF=[]
for i in relation2['Speed.value'].index:
    df = relation2['Speed.value'].at[i]
    #print(df)
    value = gaussianKernelRegression.predict(df)
    #print(i, df, value)
    predictedDF.append({'Speed.value':df, 'predicted CO2 Emission (GPS-based).value': value})
predictedDF=pd.DataFrame(predictedDF)

In [ ]:
predictedDF

In [ ]:
y = np.c_[relation2["CO2 Emission (GPS-based).value"]]
y_predicted=np.c_[predictedDF["predicted CO2 Emission (GPS-based).value"]]

rmse_n = math.sqrt(mean_squared_error(y, y_predicted))
rmse_n

As this is an error of 3.5, it turns out, that Speed value in this form may not be a good single predictor to predict CO2 Comsumption. However, this was a rough analysis without further cleaning or transforming the data which may affect the result.